File: apps/backend-python/app/modules/document/application/commands/delete_document_handler.py
"""
Delete Document Handler

Обработчик команды удаления документа.
"""
from uuid import UUID

from app.shared.domain.result import Result
from app.modules.document.application.commands.delete_document_command import (
    DeleteDocumentCommand,
)
from app.modules.document.domain.repositories.document_repository import (
    IDocumentRepository,
)


class DeleteDocumentHandler:
    """
    Handler для команды удаления документа.

    Orchestrates:
    1. Поиск документа
    2. Проверка владельца
    3. Пометка на удаление (soft delete)
    4. Сохранение изменений

    Note: Физическое удаление файла из S3 происходит асинхронно
    через обработчик события DocumentDeletedEvent.
    """

    def __init__(self, document_repository: IDocumentRepository):
        """
        Инициализирует handler.

        Args:
            document_repository: Репозиторий документов
        """
        self.document_repository = document_repository

    async def handle(self, command: DeleteDocumentCommand) -> Result[None]:
        """
        Обрабатывает команду удаления документа.

        Args:
            command: Команда удаления

        Returns:
            Result с успехом или ошибкой
        """
        # 1. Находим документ
        document_id = UUID(command.document_id)
        document = await self.document_repository.find_by_id(document_id)

        if document is None:
            return Result.fail(f"Document not found: {command.document_id}")

        # 2. Проверяем владельца
        if str(document.owner_id) != command.owner_id:
            return Result.fail(
                "Access denied. You can only delete your own documents."
            )

        # 3. Помечаем документ на удаление через доменный метод
        delete_result = document.delete()

        if not delete_result.is_success:
            return Result.fail(delete_result.error)

        # 4. Сохраняем изменения (soft delete)
        await self.document_repository.save(document)

        # 5. Domain event DocumentDeletedEvent будет обработан
        # асинхронно для физического удаления файла из S3

        return Result.ok()

File: apps/backend/src/modules/document/application/commands/delete-document/delete-document.handler.ts
import { CommandHandler, ICommandHandler } from '@nestjs/cqrs';
import { Injectable } from '@nestjs/common';
import { DeleteDocumentCommand } from './delete-document.command';
import { DocumentRepository } from '../../../domain/repositories/document.repository';
import { Result } from '../../../../../shared/domain/result';
import { IStorageService } from '../../../domain/services/storage.service';

@CommandHandler(DeleteDocumentCommand)
@Injectable()
export class DeleteDocumentHandler implements ICommandHandler<DeleteDocumentCommand, Result<void>> {
  constructor(
    private readonly documentRepository: DocumentRepository,
    private readonly storageService: IStorageService,
  ) {}

  async execute(command: DeleteDocumentCommand): Promise<Result<void>> {
    try {
      // 1. Load document
      const document = await this.documentRepository.findById(command.documentId);
      if (!document) {
        return Result.fail('Document not found');
      }

      // 2. Authorization check
      if (document.lawyerId !== command.requestingUserId) {
        return Result.fail('Unauthorized: You can only delete your own documents');
      }

      // 3. Delete chunks
      await this.documentRepository.deleteChunksByDocumentId(document.id);

      // 4. Delete document record
      await this.documentRepository.delete(document.id);

      // 5. Delete file from storage (async, don't wait)
      const fileKey = this.extractFileKeyFromUrl(document.fileUrl);
      this.storageService.deleteFile(fileKey).catch((error) => {
        // Log error but don't fail the command
        console.error(`Failed to delete file from storage: ${error.message}`);
      });

      return Result.ok();
    } catch (error) {
      return Result.fail(`Unexpected error: ${error.message}`);
    }
  }

  private extractFileKeyFromUrl(url: string): string {
    try {
      const urlObj = new URL(url);
      const pathParts = urlObj.pathname.split('/');
      const publicIndex = pathParts.indexOf('public');
      if (publicIndex !== -1) {
        return pathParts.slice(publicIndex + 1).join('/');
      }
      const objectIndex = pathParts.indexOf('object');
      if (objectIndex !== -1) {
        return pathParts.slice(objectIndex + 1).join('/');
      }
      throw new Error('Could not extract file key from URL');
    } catch (error) {
      throw new Error(`Invalid file URL: ${url}`);
    }
  }
}
